cpp/src/generators/erdos_renyi_generator.cu

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph/graph_generators.hpp>
+#include <cugraph/utilities/error.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/copy.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/random.h>
+
+namespace cugraph {
+
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+generate_erdos_renyi_graph_edgelist(raft::handle_t const& handle,
+                                    vertex_t num_vertices,
+                                    float p,
+                                    vertex_t base_vertex_id,
+                                    uint64_t seed)
+{
+  auto random_iterator = thrust::make_transform_iterator(
+    thrust::make_counting_iterator<size_t>(0), [seed] __device__(size_t index) {
+      thrust::default_random_engine rng(seed);
+      thrust::uniform_real_distribution<float> dist(0.0, 1.0);
+      rng.discard(index);
+      return dist(rng);
+    });
+
+  size_t count = thrust::count_if(rmm::exec_policy(handle.get_stream()),
+                                  random_iterator,
+                                  random_iterator + num_vertices * num_vertices,
+                                  [p] __device__(float prob) { return prob < p; });
+
+  rmm::device_uvector<size_t> indices_v(count, handle.get_stream());
+
+  handle.get_stream_view().synchronize();
+
+  thrust::copy_if(rmm::exec_policy(handle.get_stream()),
+                  random_iterator,
+                  random_iterator + num_vertices * num_vertices,
+                  indices_v.begin(),
+                  [p] __device__(float prob) { return prob < p; });
+
+  rmm::device_uvector<vertex_t> src_v(count, handle.get_stream());
+  rmm::device_uvector<vertex_t> dst_v(count, handle.get_stream());
+
+  thrust::transform(rmm::exec_policy(handle.get_stream()),
+                    indices_v.begin(),
+                    indices_v.end(),
+                    thrust::make_zip_iterator(thrust::make_tuple(src_v.begin(), src_v.end())),
+                    [num_vertices] __device__(size_t index) {
+                      size_t src = index / num_vertices;
+                      size_t dst = index % num_vertices;
```

Comment:
This will break if vertex_t is 64 bit.

Reply:
Technically, it will break if num_vertices > (2^31 - 1).  If vertex_t is 64 bits but could be stored in 32 bits it would still work.

Since we don't want this variation called if num_vertices > (2^31 - 1)... even if it worked we don't want to do that much computation... I just added a CUGRAPH_EXPECTS at the beginning of the function.